Add DoubleSelectiveStreamReader

sayhar opened the pull request:
Fixes #13191

```
== NO RELEASE NOTE ==
````

sayhar commented:
Unit and integration test pass
No more inlining of function

My best guess (okay, hope) is that you'll give a bunch of nits/comments around code style, I'll fix them, and then this will be ready

bhhari commented:
@sayhar can you squash the commits into one.

sayhar commented:
Squashed

bhhari commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -132,6 +132,30 @@ public void testNumeric()
+    @Test
+    public void testDouble()
+    {
+        assertQuery("SELECT quantity, extendedprice, discount, tax from lineitem");
```
you should use assertQueryUsingH2Cte as we are comparing the hive read with the h2 read.

sayhar replied:
Like this? 
```
        // Nulls
        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE ship_day is null");

        assertQueryUsingH2Cte("SELECT sum(quantity) FROM lineitem_ex WHERE ship_by_air is null");

        assertQueryUsingH2Cte("SELECT sum(quantity) FROM lineitem_ex WHERE ship_by_air is null");

        assertQueryUsingH2Cte("SELECT quantity * (1 - discount) * (1 + tax) FROM lineitem_ex WHERE is_returned = true");
```

bhhari replied:
You need to also generate null data for the queries to work. similar to how it was generated for Boolean and Byte readers
```
CASE WHEN linenumber % 5 = 0 THEN null ELSE shipmode = 'AIR' END AS ship_by_air
```

sayhar replied:
Resolved, right?

bhhari commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,430 @@
+    private boolean allNulls;
+    private boolean valuesInUse;
+
+    private LocalMemoryContext systemMemoryContext;
```
nit: make this final and move it next to other final fields

sayhar replied:
Got it

bhhari commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java:
```diff
@@ -52,6 +52,8 @@ public static SelectiveStreamReader createStreamReader(
                 return new LongSelectiveStreamReader(streamDescriptor, filter, outputType, systemMemoryContext);
             case FLOAT:
```
add a throw here or break, as even FLOAT now falls through DOUBLE

sayhar replied:
good call

sayhar replied:
resolved?

bhhari commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,430 @@
+public class DoubleSelectiveStreamReader
+        implements SelectiveStreamReader
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BooleanSelectiveStreamReader.class).instanceSize();
```
change the class to DoubleSelectiveStreamReader

sayhar replied:
Good catch! Sad I missed this one

sayhar commented:
@bhhari  I think I fixed all your nits. Let me know!

bhhari commented:
looks good, only 1 change needed, fix the test TestHivePushDownFilters to generate null values for both h2 and hive and also put the Benchmark results in the commit summary.

mbasmanova commented:
@sayhar This change has no user visible effects, hence, no need for release notes.

Travis reports stylecheck failures:

```
[INFO] Checking compile classpath
[INFO] There are 8 errors reported by Checkstyle 8.7 with src/checkstyle/presto-checks.xml ruleset.
[ERROR] src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:[29] (imports) ImportOrder: 'java.lang.Double.doubleToLongBits' should be separated from previous imports.
[ERROR] src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:[30] (imports) ImportOrder: Wrong order for 'org.openjdk.jol.info.ClassLayout' import.
[ERROR] src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:[80] (regexp) RegexpMultiline: Multiple consecutive blank lines
[ERROR] src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:[186] (regexp) RegexpMultiline: Blank line after opening brace
[ERROR] src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:[68] (imports) ImportOrder: Wrong order for 'com.facebook.presto.spi.type.DoubleType.DOUBLE' import.
[ERROR] src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:[258,32] (whitespace) WhitespaceAround: '{' is not preceded with whitespace.
[ERROR] src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:[300,15] (whitespace) WhitespaceAfter: 'if' is not followed by whitespace.
[ERROR] src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:[47] (imports) ImportOrder: Wrong order for 'com.facebook.presto.spi.type.DoubleType.DOUBLE' import.
```

You can reproduce these locally by running `mvn install -DskipTests -T1C -pl '!presto-docs,!presto-server-rpm'` command.

sayhar commented:
Wait, so do I add benchmark results in the release notes or not?

Here they are just for ease of finding next time: 
```
// With inlining

# Run complete. Total time: 00:04:48

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.172 ±  0.010   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.280 ±  0.015   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.182 ±  0.006   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.324 ±  0.012   s/op


// With function call


Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.176 ±  0.017   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.271 ±  0.005   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.162 ±  0.009   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.318 ±  0.012   s/op
```

bhhari commented:
> Wait, so do I add benchmark results in the release notes or not?
> 
> Here they are just for ease of finding next time:
> 
> ```
> // With inlining
> 
> # Run complete. Total time: 00:04:48
> 
> Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
> BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.172 ±  0.010   s/op
> BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.280 ±  0.015   s/op
> BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
> BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.182 ±  0.006   s/op
> BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.324 ±  0.012   s/op
> 
> 
> // With function call
> 
> 
> Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
> BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.176 ±  0.017   s/op
> BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.271 ±  0.005   s/op
> BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
> BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.162 ±  0.009   s/op
> BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.318 ±  0.012   s/op
> ```

yes please add the Benchmark for the Selective reader in the commit, take a look at https://github.com/prestodb/presto/pull/13193/commits, when you expand the "..." you will see the benchmark results..something that @mbasmanova mentioned, so that we can look back on the results during the commit

sayhar commented:
@bhhari  like so?

bhhari commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -37,7 +37,7 @@
 
     private static final String WITH_LINEITEM_EX = "WITH lineitem_ex AS (\n" +
-            "SELECT linenumber, orderkey, partkey, suppkey, \n" +
+            "SELECT linenumber, orderkey, partkey, suppkey, quantity, discount, tax, \n" +
```
use some condition similar to to generate some null values
```
CASE WHEN linenumber % 5 = 0 THEN null ELSE discount
```

bhhari commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -62,8 +62,8 @@ private static QueryRunner createQueryRunner()
         queryRunner.execute(noPushdownFilter(queryRunner.getDefaultSession()),
-                "CREATE TABLE lineitem_ex (linenumber, orderkey, partkey, suppkey, ship_by_air, is_returned, ship_day, ship_month, keys, nested_keys, flags) AS " +
-                        "SELECT linenumber, orderkey, partkey, suppkey, " +
+                "CREATE TABLE lineitem_ex (linenumber, orderkey, partkey, suppkey, quantity, discount, tax, ship_by_air, is_returned, ship_day, ship_month, keys, nested_keys, flags) AS " +
```
same as above, generate some null values

mbasmanova commented:
@sayhar 

> Wait, so do I add benchmark results in the release notes or not?

No. Add them to the commit message. BTW, how do these results compare to batch reader?

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -308,4 +329,15 @@ protected T computeNext()
             }
         };
     }
+    // Copied from AbstractTestOrcReader:
```
- this comment is not needed
- add an empty line before this method

mbasmanova commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -308,4 +329,15 @@ protected T computeNext()
+    // Copied from AbstractTestOrcReader:
+    private static List<Double> doubleSequence(double start, double step, int items)
+    {
+        List<Double> values = new ArrayList<>();
```
nit: this can be shortened to 

```
        return IntStream.range(0, items)
                .mapToDouble(i -> start + i * step)
                .boxed()
                .collect(toImmutableList());
```

sayhar commented:
> No. Add them to the commit message.

They should be added to the commit message. I'll edit the github comment to make that clearer.
>  BTW, how do these results compare to batch reader?

Here's the raw results:
Seems like reading no nulls, no filter is roughly comparable, with an edge to batch.
No nulls, with filter gives a clear win to *batch*. (!?!?!?)
With nulls and filters, selective reader does better. 

But perhaps I'm misreading. 

```

// With function call

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.180 ±  0.018   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.290 ±  0.020   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.174 ±  0.005   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.324 ±  0.014   s/op


// batch version

BenchmarkBatchStreamReaders.readDoubleNoNull                         N/A  avgt   60  0.278 ±  0.018   s/op
BenchmarkBatchStreamReaders.readDoubleWithNull                       N/A  avgt   60  0.215 ±  0.006   s/op
```

mbasmanova commented:
@sayhar 

> Seems like reading no nulls, no filter is roughly comparable, with an edge to batch.

The errors are pretty high (~10%). I'm thinking that we need to adjust the benchmark to lower the errors and then measure again.

> No nulls, with filter gives a clear win to batch. (!?!?!?)

I think I have an explanation. The filter allows any values from -infinity to +infinity, hence, the reader materializes all the values and spends extra cycles on evaluating the filter. Batch reader just materializes all the values. I'd change the filter to select only half of the rows, e.g. [0, +infinity).

bhhari commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,432 @@
+            }
+
+            if (presentStream != null && !presentStream.nextBit()) {
+                if (nullsAllowed) {
```
check for the nonDeterministicFilter here

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,432 @@
+            skip(offset - readOffset);
+        }
+
+        int streamPosition = 0;
```
Aha! I knew it! 
(Should have gone with my instinct when copying this code)

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,432 @@
+        }
+
+        if (filter != null) {
+            ensureOutputPositionsCapacity(positionCount);
```
Seems like a fine idea to me. 
In fact, while ensureValuesCapacity seems to perhaps merit keeping as a method (but using Arrays.ensureCapacity internally), why bother keeping ensureOutputPositionsCapacity at all? I propose removing it as a function and straight up replacing line 165 with 
`outputPositions = Arrays.ensureCapacity(outputPositions, positionCount);`

mbasmanova replied:
Yes. ensureOutputPositionsCapacity can be replaced with ensureCapacity:

`outputPositions = ensureCapacity(outputPositions, positionCount);`

note: static import Arrays.ensureCapacity

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:
```diff
@@ -251,6 +255,10 @@ public void setup()
             }
 
+            if (type == DOUBLE) {
+                return Optional.of(DoubleRange.of(Double.MIN_VALUE, false, false, Double.MAX_VALUE, false, false, false));
```
Not intentional, you're right, and I've fixed this in my upcoming amend

sayhar commented:
@mbasmanova 
> I think I have an explanation. The filter allows any values from -infinity to +infinity, hence, the reader materializes all the values and spends extra cycles on evaluating the filter. Batch reader just materializes all the values. I'd change the filter to select only half of the rows, e.g. [0, +infinity).

Of course!

I've changed to 50k rows in the benchmark file, and run 3 different versions of benchmarks:
filter from 
`0 -> Double.MAX_VALUE / 2` 
to 
`0 -> Double.MAX_VALUE / 20`
to
`Double.MAX_VALUE / 2000`.

The benchmark results seem to quite similar for each version. The error bars still seem high. How do you suggest tweaking this beyond increasing the row count? (Or perhaps I should increase it still higher, from 50k to maybe 200k?)

Here are the results:

```
// With function (50k rows, 0 -> max / 2)
# Run complete. Total time: 00:08:14

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.913 ± 0.072   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  1.389 ± 0.040   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.014 ± 0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.923 ± 0.070   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  1.596 ± 0.062   s/op
```
```
// With function (50k rows, 0 -> max / 20)

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.914 ± 0.066   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  1.346 ± 0.055   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.014 ± 0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.926 ± 0.053   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  1.557 ± 0.062   s/op
```
```
// With function (50k rows, 0 -> max / 2000)
# Run complete. Total time: 00:08:13

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score   Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.924 ± 0.090   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  1.393 ± 0.059   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.015 ± 0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  1.049 ± 0.132   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  1.556 ± 0.032   s/op
```

sayhar commented:
@bhhari  @mbasmanova -- I think I've addressed all your comments except one: 
(This one:)
> This logic is repeated in Timestamp reader. Consider extracting it into a helper class. CC: 

I have a few questions that came up in the course of making these changes. 

They are as follows:
```
- Lines 201 - 204 in BooleanSelectiveStreamReader: 
	- Why this logic?
	- `(nonDeterministicFilter && filter.testNull()) || nullsAllowed)` this will always return true
- ReadAllNulls in line 197 of DoubleSelectiveStreamReader:
	- What does this actually do?
	- I think I know the answer
	- There's a difference between returning an array of nulls and an empty array (of size 0)
		- (A big difference! The difference between a Where clause filtering out a row or not)
		- This tries to figure out which of these it is
		- The nondeterministic filter part makes at first makes sense -- you have to figure out which to filter.
			- (Though how? The only way this is invoked is if the datastream is null. And if all the data is null, how can you figure out which to filter and which not to?)
			- I saw in the documentation that nonDeterministicFilters has to do with things like array lookups? But how does that fit in with no dataStream?
- Why the name "TupleDomainFilter?" What is a domain in this context?
- Line ~165 of DoubleSelectiveStreamReader
	-  Re "Would you use Arrays.ensureCapacity methods here and in ensureValuesCapacity?"
		- Yeah, good idea.
		- Which is better? Keeping ensureOutputPositionsCapacity and changing its internals to use Arrays.ensureCapacity, or just removing it as a method altogether?
- TestHivePushdownFilterQueries
	- This is what I'm using for integration tests
	- But how can I see that its using selective filters and not the legacy batch filters?

```

sayhar commented:
Also: If I understand correctly, benchmarking batch and selective stream readers use different methods to create different sets of test data. 

Given that, can we credibly compare the apples and oranges of the two different benchmarks?

mbasmanova commented:
@sayhar 

> Also: If I understand correctly, benchmarking batch and selective stream readers use different methods to create different sets of test data.

> Given that, can we credibly compare the apples and oranges of the two different benchmarks?

The code looks different, but actual data sets are the same. This is confusing though. It would be helpful to refactor batch benchmark to match selective benchmark.

mbasmanova commented:
@sayhar 

> I have a few questions that came up in the course of making these changes.

That's a nice set of questions. Let me try to answer them.

> - Lines 201 - 204 in BooleanSelectiveStreamReader: 
> 	- Why this logic?
> 	- `(nonDeterministicFilter && filter.testNull()) || nullsAllowed)` this will always return true

Non-deterministic filters occur when query applies a range filter to a nested column of an array, a map or a struct, e.g. `a[2] > 10` or `a[2] = true` or `a[2] is null`. In this case, the filters `> 10`, `= true` and `is null` apply to some rows, but not others. Hence, it is non-deterministic. Check out #13099 for further explanation and let me know if you have follow up questions.

> - ReadAllNulls in line 197 of DoubleSelectiveStreamReader: ... But how does that fit in with no dataStream?

> - Why the name "TupleDomainFilter?" What is a domain in this context?

See TupleDomain class: `Defines a set of valid tuples according to the constraints on each of its constituent columns`. TupleDomain represents simple/range filters as a set of valid ranges (domain). `a between 10 and 20` becomes a single range [10, 20]. `a in (2, 3, 5)` becomes a set of ranges {[2-3], [5]}. `a not in (2, 3, 5)` becomes {(-inf, 1], [4], [6, +inf)}. TupleDomainFilter is a filter that corresponds to a TupleDomain.

> - TestHivePushdownFilterQueries
> 	- This is what I'm using for integration tests
>	- But how can I see that its using selective filters and not the legacy batch filters?

L61 `ImmutableMap.of("hive.pushdown-filter-enabled", "true"),` sets configuration flag that enables optimized scan path. You can replace the code that created double reader in SelectiveStreamReaders#createStreamReader with `throw new IllegalArgumentException("Unsupported type: " + type);` and re-run the test. It will fail and the stack trace will point to that line.

bhhari commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReaders.java:
```diff
@@ -280,21 +280,22 @@ OrcBatchRecordReader createRecordReader()
-
-                "varchar",
-                "varbinary",
+//                "boolean",
```
you will need to uncomment these

sayhar replied:
Sorry, forgot to push the change to this

sayhar commented:
@mbasmanova
> . In this case, the filters > 10, = true and is null apply to some rows, but not others. Hence, it is non-deterministic. 

I mean very mechanistically: If `nonDeterministicFilter` is true, then `nullsAllowed` is true, and you don't need to check `filter.testNull()`.

This is true because the expression `nullsAllowed = this.filter == null || nonDeterministicFilter || this.filter.testNull();` on line 97.

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,432 @@
+            return block;
+        }
+
+        long[] valuesCopy = new long[positionCount];
```
> This logic is repeated in Timestamp reader. Consider extracting it into a helper class. CC: @bhhari

Can we make this a different issue?

sayhar commented:
@mbasmanova  @bhhari  I think I've resolved all the issues you've brought up, except @mbasmanova's idea of abstracting some duplicated code (which I propose should be a different issue for commit separability reasons. 

I've also compared benchmarks for selective vs legacy/batch stream reader. The results are:
- Reading with no filter is always faster with selective reader
- Reading with a filter and with nulls is always faster with selective reader
- Reading with a filter and no nulls is faster with batch reader 
- No clear pattern between a more restrictive and less restrictive filter. 

Do these results match our expectations?
- I think it makes sense that a selectivestreamreader would take longer to read with a filter than a batch reader with no filter -- it has to do one more set of operations (comparison). Its only when chaining a few selectivereaders that we should see benchmarking gains. Right?

```
// standard batch benchmark
// 10 million rows
// reordered for easier comparison
Benchmark                                                (typeSignature)  Mode  Cnt  Score    Error  Units
BenchmarkBatchStreamReaders.readDoubleWithNull                       N/A  avgt   60  0.205 ±  0.005   s/op
BenchmarkBatchStreamReaders.readDoubleNoNull                         N/A  avgt   60  0.271 ±  0.007   s/op
```

```
// Selective filter = Optional.of(DoubleRange.of(.9, false, false, 2, false, false, false));
// 10 million

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.176 ±  0.009   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.265 ±  0.011   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.167 ±  0.004   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.271 ±  0.003   s/op
```

```
// Selective Filter =  (DoubleRange.of(.5, false, false, 2, false, false, false));
// 10 million rows

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.151 ±  0.008   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.255 ±  0.004   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.193 ±  0.008   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.336 ±  0.012   s/op
```

```
// Selective filter = Optional.of(DoubleRange.of(.1, false, false, 2, false, false, false));
// 10 million rows

Benchmark                                       (typeSignature)  (withNulls)  Mode  Cnt  Score    Error  Units
BenchmarkSelectiveStreamReaders.read                     double         true  avgt   20  0.154 ±  0.009   s/op
BenchmarkSelectiveStreamReaders.read                     double        false  avgt   20  0.259 ±  0.003   s/op
BenchmarkSelectiveStreamReaders.readAllNull              double          N/A  avgt   20  0.003 ±  0.001   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double         true  avgt   20  0.174 ±  0.006   s/op
BenchmarkSelectiveStreamReaders.readWithFilter           double        false  avgt   20  0.329 ±  0.021   s/op
```

sayhar commented:
Running tests `(com/facebook/presto/hive/TestHivePushdownFilterQueries.java:175/testDouble())`
and getting errors that are clearly not from the work I'm doing. 

Like: 
```
/Users/red/presto/presto-hive/src/main/java/com/facebook/presto/hive/gcs/HiveGcsConfigurationInitializer.java
Error:(16, 38) java: package com.google.cloud.hadoop.fs.gcs does not exist
```

Did something happen upstream? Why hasn't merging upstream fixed these changes?

mbasmanova commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -169,6 +171,38 @@ public void testTimestamps()
         assertQueryUsingH2Cte("SELECT commit_timestamp, ship_timestamp, orderkey FROM lineitem_ex WHERE year(commit_timestamp) > 1993 and year(ship_timestamp) > 1993 and year(ship_timestamp) - year(commit_timestamp) = 1");
     }
+    @Test
+    public void testDouble()
```
It looks like `compactValues` code path is not tested. Overall, we need to test the following cases:

- read all or selected positions
- no filter or with filter
- project data out or filter only
- apply filter function (e.g. call getBlockView and call getBlock with a subset of positions)

assertFilterProject helps cover "project data out or filter only", hence, I recommend using it.

Here are some examples of the queries and coverage they provide.

assertFilterProject("discount = 4", "tax"); // tax reader reads selected positions with no filter; discount reader reads all positions with filter

assertFilterProject("discount > tax", "tax"); // discount and tax readers read all positions with no filter; getBlockView is called for discount and tax; getBlock is called with a subset of positions

assertFilterProject("discount > tax and tax < 0.01", "tax"); // tax reader reads all with filter; discount reader reads selected positions with no filter; getBlockView is called for discount and tax; compactValues is called for discount; getBlock is called with a subset of positions

sayhar replied:
Trying to find a way to trigger the compact values code path. 

For example: 
```
assertFilterProject("discount < tax AND (extendedprice > 1500 * quantity OR discount * 100 > quantity)", "extendedprice");
```

Doesn't trigger `compactValues`. 
```
SELECT count(*) 
FROM  tpch.tiny.lineitem
WHERE 
discount < tax AND (extendedprice > 1500 * quantity OR discount * 100 > quantity)
```
returns 9121 rows, so you'd think compacting would be used. Help!

Similarly, `readAllNulls` is never called. I imagine because we never write test data that has a `presentStream` and no `dataStream`?

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -264,62 +365,4 @@ private void testRoundTripNumeric(Iterable<? extends Number> values, TupleDomain
                         ImmutableMap.of(0, filter, 1, filter, 2, filter)));
     }
-
-    private static ContiguousSet<Integer> intsBetween(int lowerInclusive, int upperExclusive)
```
Moved the method higher up in the file to match Presto code formatting

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkBatchStreamReaders.java:
```diff
@@ -281,18 +281,14 @@ OrcBatchRecordReader createRecordReader()
         @SuppressWarnings("unused")
         @Param({
                 "boolean",
-
```
Maybe. I'm just clearing blank spaces here, aren't I? Why the blank spaces in the first place?

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkSelectiveStreamReaders.java:
```diff
@@ -88,6 +90,34 @@
     public static final int ROWS = 10_000_000;
     public static final List<?> NULL_VALUES = Collections.nCopies(ROWS, null);
 
+    private static List<Block> readAllBlocks(OrcSelectiveRecordReader recordReader)
```
Auto-Formatting (I thought the convention was to autoformat aggressively, all the time)

sayhar commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHivePushdownFilterQueries.java:
```diff
@@ -169,6 +171,38 @@ public void testTimestamps()
+    public void testDouble()
+    {
+        // Nulls
+        assertQueryUsingH2Cte("SELECT count(*) FROM lineitem_ex WHERE ship_day is null");
```
I wanted to have a gradually increasing amount of complexity per test, so I wanted to start from the most basic. That said, I think ship_day should have been `discount` or another double instead

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/SelectiveStreamReaders.java:
```diff
@@ -61,7 +61,10 @@ public static SelectiveStreamReader createStreamReader(
             case FLOAT:
+                throw new IllegalArgumentException("Unsupported type: " + streamDescriptor.getStreamType());
             case DOUBLE:
+                checkArgument(requiredSubfields.isEmpty(), "Double stream reader doesn't support subfields");
```
Also -- on line 114 / 111, shouldn't there be a change there too? I just noticed it.

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -146,6 +214,39 @@ public void testByteValuesPatchedBase()
+    public void testDoubleSequence()
+            throws Exception
+    {
+        tester.testRoundTrip(DOUBLE, doubleSequence(0, 0.1, 30_000, false));
```
Okay removing 221 and 222

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -146,6 +214,39 @@ public void testByteValuesPatchedBase()
+        tester.testRoundTrip(DOUBLE, doubleSequence(0, 0.1, 30_000, false));
+        tester.testRoundTrip(DOUBLE, doubleSequence(0, 0.1, 30_000, true));
+
+        List<Map<Integer, TupleDomainFilter>> filters = ImmutableList.of(
```
You're right -- these filters are left over from when I had an incomplete understanding of the data we are testing. When I updated my understanding I should have updated this line too. 

Your list of filters seems great, I'll use that.

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -146,6 +214,39 @@ public void testByteValuesPatchedBase()
+                ImmutableMap.of(0, DoubleRange.of(1, false, false, 4, false, true, true)),
+                ImmutableMap.of(0, DoubleRange.of(0, false, false, 40, false, false, true)));
+
+        tester.testRoundTrip(DOUBLE, doubleSequence(0, 0.1, 30_000, false), filters);
```
I see! Nice!

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -146,6 +214,39 @@ public void testByteValuesPatchedBase()
+    public void testDoubleNaNInfinity()
+            throws Exception
+    {
+        tester.testRoundTrip(DOUBLE, ImmutableList.of(1000.0, -1.0, Double.POSITIVE_INFINITY));
```
good idea

sayhar commented on presto-orc/src/main/java/com/facebook/presto/orc/reader/DoubleSelectiveStreamReader.java:
```diff
@@ -0,0 +1,430 @@
+                    outputPositionCount++;
+                }
+            }
+            streamPosition++;
```
I don't understand. 
Will come back to this

sayhar commented on presto-orc/src/test/java/com/facebook/presto/orc/TestSelectiveOrcReader.java:
```diff
@@ -61,6 +63,72 @@
+        };
+    }
+
+    private static List<Double> doubleSequence(double start, double step, int items, boolean hasNulls)
```
I'll move it to the bottom of the file. Or did you mean *right* after the function that calls it?

sayhar commented:
Just pushed a new version. 
It should have addressed all comments (including re-ordering method order) except: 

### 1 
https://github.com/prestodb/presto/pull/13208#discussion_r316492933
> In  case of a non-deterministic positional filter, it is necessary to check for filter.getPrecedingPositionsToFail() and filter.getSucceedingPositionsToFail();

and this:
### 2 
https://github.com/prestodb/presto/pull/13208#discussion_r316499509
> It looks like compactValues code path is not tested.

and I noticed this:

### 3.
https://github.com/prestodb/presto/pull/13208#discussion_r316738990
> on line 114 / 111 [on SelectiveStreamerReaders.java], shouldn't there be a change there too? 


For (1): I just don't understand. 
For (2): I haven't been able to trigger that code path. 
For (3): I noticed this myself, flagging for others

bhhari commented:
For (1): I just don't understand.
This is when you have array(double) filter, which @mbasmanova  recently implemented in ListSelectiveStreamReader.
If you have a column array(double, int, string) and sql query with those
select count(*) where arr[1] = 1.0 and a[2] = 1

For (2): I haven't been able to trigger that code path.
This path will be triggered with the TestHivePushDownQueries where you have a function using 2 columns select a,b where b>5 and a >10 and a+b > 20

For (3): I noticed this myself, flagging for others
that code gets triggered only when the list support is added for those scalars

sayhar commented:
Thanks to @bhhari  I fixed (2) 

There are a few code paths still not covered. Will follow up tomorrow

sayhar commented:
Thanks to talking to @bhhari I think I now understand #1 conceptually

sayhar commented:
All code in DoubleSelectStream reader is covered except lines:

```
203 - 207 (waiting on structs, see comment)
308  (nulls)
393  (nulls)
428  (nulls)
```

Note that I added a few more flags to TestSelectiveOrcReader.java

bhhari commented:
these properties are resulting in test failures

 orcTester.reverseTestsEnabled = true; ( does not work with null values) this can be fixed by changing ImmutalbeList to a Lists

```
private static <T> List<T> reverse(List<T> iterable)
    {
        return Lists.reverse(Lists.newArrayList(iterable));
    }

```
 orcTester.missingStructFieldsTestsEnabled = true;
 orcTester.skipStripeTestsEnabled = true;

sayhar commented:
I'll just not add the properties for now
